apps/test/authoredHintUtilsTest.js

```diff
@@ -0,0 +1,255 @@
+      authoredHintUtils.recordUnfinishedHint(hintOne);
+      assert.equal(localStorage.getItem('unfinished_authored_hint_views'), JSON.stringify([hintOne]));
+      authoredHintUtils.finishHints(record);
+      var finishedHints = authoredHintUtils.getFinishedHints_();
+
+      assert.deepEqual(finishedHints, [{
+        id: 'first',
+        prev_time: 'something decent',
+        next_time: 'something great',
+        final_time: 'something awesome',
+        next_test_result: 'something grand',
+        next_activity_id: 0,
+        next_attempt: 1,
+        next_level_source_id: 2
+      }]);
+    });
+
+  });
+
+  describe('submitHints', function () {
+
+    it('finishes all unfinished hints using the last hint', function () {
+      $.ajax = function () {};
+
+      localStorage.setItem('last_attempt_record', JSON.stringify(record));
+      authoredHintUtils.recordUnfinishedHint(hintOne);
+
+      localStorage.setItem('last_attempt_record', JSON.stringify(recordTwo));
+      authoredHintUtils.recordUnfinishedHint(hintTwo);
+
+      authoredHintUtils.submitHints('/url');
+
+      var finishedHints = authoredHintUtils.getFinishedHints_();
+
+      assert.deepEqual(finishedHints, [{
+        id: 'first',
+        final_time: 'something awesome',
+        next_activity_id: 3,
+        next_attempt: 4,
+        next_level_source_id: 5,
+        next_test_result: 'something okay',
+        next_time: 'something great',
+        prev_activity_id: 0,
+        prev_attempt: 1,
+        prev_level_source_id: 2,
+        prev_test_result: 'something grand',
+        prev_time: 'something decent',
+      }, {
+        id: 'second',
+        final_test_result: 'something rather poor',
+        next_activity_id: 3,
+        next_attempt: 4,
+        next_level_source_id: 5,
+        next_test_result: 'something fine',
+        next_time: 'something superb',
+        prev_activity_id: 3,
+        prev_attempt: 4,
+        prev_level_source_id: 5,
+        prev_test_result: 'something okay',
+        prev_time: 'something superb',
+      }]);
+
+    });
+
+    it('posts all finalized hints at once', function () {
+      var data;
+      $.ajax = function (options) {
+        data = options.data;
+      };
+      var hints = [1, 2, 3];
+      localStorage.setItem('finished_authored_hint_views', JSON.stringify(hints));
+      var finalizedHints = authoredHintUtils.finalizeHints_();
+
+      authoredHintUtils.submitHints('/url');
+
+      assert.deepEqual(data, JSON.stringify({
+        hints: finalizedHints
+      }));
+    });
+
+    it('clears finished hints after successful post', function () {
+      $.ajax = function (options) {
+        options.complete();
+      };
+      authoredHintUtils.recordUnfinishedHint(hintOne);
+      authoredHintUtils.finishHints(record);
+      authoredHintUtils.submitHints('/url');
+      var finishedHints = authoredHintUtils.getFinishedHints_();
+      assert.deepEqual(finishedHints, []);
+    });
+
+  });
+
+});
```

Comment:
Thanks! I had to find some way to assuage my guilt for designing such a complex process